iothub/device/src/Transport/Amqp/AmqpUnit.cs

```diff
@@ -82,91 +73,145 @@ public async Task OpenAsync(TimeSpan timeout)
+                    _messageSendingLink.Closed += (obj, arg) => {
+                        _amqpIoTSession.SafeClose();
+                    };
+
+                    if (Logging.IsEnabled) Logging.Associate(this, _messageSendingLink, $"{nameof(_messageSendingLink)}");
                 }
-
-                throw;
+                return _amqpIoTSession;
             }
             finally
             {
-                if (Logging.IsEnabled) Logging.Exit(this, timeout, $"{nameof(OpenAsync)}");
+                if (Logging.IsEnabled) Logging.Exit(this, timeout, $"{nameof(EnsureSessionAsync)}");
+                _sessionLock.Release();
             }
         }
 
         public async Task CloseAsync(TimeSpan timeout)
         {
             if (Logging.IsEnabled) Logging.Enter(this, timeout, $"{nameof(CloseAsync)}");
+            
+            bool gain = await _sessionLock.WaitAsync(timeout).ConfigureAwait(false);
+            if (!gain)
+            {
+                throw new TimeoutException();
+            }
 
-            if (SetNotUsable() == 0 && _amqpIoTSession != null)
+            try
             {
-                await _amqpIoTSession.CloseAsync(timeout).ConfigureAwait(false);
-                OnUnitDisconnected?.Invoke(true, EventArgs.Empty);
+                if (_amqpIoTSession != null && !_amqpIoTSession.IsClosing())
+                {
+                    try
+                    {
+                        await _amqpIoTSession.CloseAsync(timeout).ConfigureAwait(false);
+                    }
+                    finally
+                    {
+                        _amqpAuthenticationRefresher?.StopLoop();
+                        _amqpIoTSession.SafeClose();
+                        if (!_deviceIdentity.IsPooling())
+                        {
+                            _amqpConnectionHolder.Dispose();
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                _closed = true;
+                if (Logging.IsEnabled) Logging.Exit(this, timeout, $"{nameof(CloseAsync)}");
+                _sessionLock.Release();
             }
 
-            if (Logging.IsEnabled) Logging.Exit(this, timeout, $"{nameof(CloseAsync)}");
         }
         #endregion
 
         #region Message
 
-        private async Task EnsureReceivingLinkIsOpenedAsync(TimeSpan timeout)
+        private async Task EnsureReceivingLinkAsync(TimeSpan timeout)
         {
-            if (Volatile.Read(ref _messageReceivingLink) != null) return;
+            if (_closed)
+            {
+                throw DEVICE_DISCONNECTED_EXCEPTION;
+            }
 
-            if (Logging.IsEnabled) Logging.Enter(this, timeout, $"{nameof(EnsureReceivingLinkIsOpenedAsync)}");
+            if (Logging.IsEnabled) Logging.Enter(this, timeout, $"{nameof(EnsureReceivingLinkAsync)}");
+            AmqpIoTSession amqpIoTSession = await EnsureSessionAsync(timeout).ConfigureAwait(false);
```

Comment:
As we have discussed yesterday, this can result in a leak of the session:

| Thread 1 | Thread 2 |
|---|---|
| SendAsync |  |
| | Dispose  (nothing cleaned since no session, link) | 
| EnsureOpened |  object marked as disposed |
| creates session |  |
| creates links |  |

End result: session and links are leaked.
This is made worse by the fact that even when Disposed or during retries, the handlers are not disposing the transport anymore. This means that the session and links are leaking and will likely hang around pinned by the Amqp library until the process ends.

Reply:
Table P0